### ISSUE TITLE
Allow fetching url parameter that includes "?" 

### ISSUE THREAD

**Issue by tmccombs** [haproxy/haproxy#1697]
### Your Feature Request

Currently, the `urlp` fetch "extracts the first occurrence of the parameter <name> in the query string, which begins after either '?' or <delim>, and which ends before '&', ';' or <delim>."

The wording in the documentation isn't completely clear, but from looking at the code, it seems that basically "?" is the default value for the delimiter, so the parameter ends if it sees another "?" after the beginning of the query string. 

For example, for a request to `https://example.com/?x=y?z`, `urlp(x)` will return `y` instead of `y?z`. This is contrary to how parsing url parameters usually works. For example running `new URL('https://example.com?x=y?z').searchParams.get('x')` in my browser's js console gives me "y?z".  I'd like something like `urlp` except without the "or <delim>" at the end of the quoted sentence.

I'm filing this as a feature request rather than a bug because `urlp` does behave as described, even if I'm not really sure when it would be desirable to work that way.

### What are you trying to do?

I have a request where one of the query parameters is itself a percent-encoded url, and haproxy uses that query parameter to generate a redirect (don't worry, it isn't an open redirect).  Unfortunately, if the url includs a query string, then the query doesn't get included in the `urlp` lookup, because of the `?`.  And while the code that generates the url is url-encoding the query parameter, the library that is doing that encoding doesn't encode the "?" because it expects that to be a valid character in a query string. 

### Output of `haproxy -vv`

```plain
HAProxy version 2.4.16-9d532c4 2022/04/29 - https://haproxy.org/
Status: long-term supported branch - will stop receiving fixes around Q2 2026.
Known bugs: http://www.haproxy.org/bugs/bugs-2.4.16.html
Running on: Linux 5.13.0-41-generic #46~20.04.1-Ubuntu SMP Wed Apr 20 13:16:21 UTC 2022 x86_64
Build options :
  TARGET  = linux-glibc
  CPU     = generic
  CC      = cc
  CFLAGS  = -O2 -g -Wall -Wextra -Wdeclaration-after-statement -fwrapv -Wno-address-of-packed-member -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-clobbered -Wno-missing-field-initializers -Wno-cast-function-type -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_PCRE2_JIT=1 USE_LIBCRYPT=1 USE_OPENSSL=1 USE_LUA=1 USE_ZLIB=1 USE_SYSTEMD=1 USE_PROMEX=1
  DEBUG   = 

Feature list : +EPOLL -KQUEUE +NETFILTER -PCRE -PCRE_JIT -PCRE2 +PCRE2_JIT +POLL -PRIVATE_CACHE +THREAD -PTHREAD_PSHARED +BACKTRACE -STATIC_PCRE -STATIC_PCRE2 +TPROXY +LINUX_TPROXY +LINUX_SPLICE +LIBCRYPT +CRYPT_H +GETADDRINFO +OPENSSL +LUA +FUTEX +ACCEPT4 -CLOSEFROM +ZLIB -SLZ +CPU_AFFINITY +TFO +NS +DL +RT -DEVICEATLAS -51DEGREES -WURFL +SYSTEMD -OBSOLETE_LINKER +PRCTL -PROCCTL +THREAD_DUMP -EVPORTS -OT -QUIC +PROMEX -MEMORY_PROFILING

Default settings :
  bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with multi-threading support (MAX_THREADS=64, default=12).
Built with OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
Running on OpenSSL version : OpenSSL 1.1.1f  31 Mar 2020
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2 TLSv1.3
Built with Lua version : Lua 5.3.3
Built with the Prometheus exporter as a service
Built with network namespace support.
Built with zlib version : 1.2.11
Running on zlib version : 1.2.11
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with PCRE2 version : 10.34 2019-11-21
PCRE2 library supports JIT : yes
Encrypted password support via crypt(3): yes
Built with gcc compiler version 9.4.0

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTTP       side=FE|BE     mux=H2       flags=HTX|CLEAN_ABRT|HOL_RISK|NO_UPG
            fcgi : mode=HTTP       side=BE        mux=FCGI     flags=HTX|HOL_RISK|NO_UPG
       <default> : mode=HTTP       side=FE|BE     mux=H1       flags=HTX
              h1 : mode=HTTP       side=FE|BE     mux=H1       flags=HTX|NO_UPG
       <default> : mode=TCP        side=FE|BE     mux=PASS     flags=
            none : mode=TCP        side=FE|BE     mux=PASS     flags=NO_UPG

Available services : prometheus-exporter
Available filters :
	[SPOE] spoe
	[CACHE] cache
	[FCGI] fcgi-app
	[COMP] compression
	[TRACE] trace
```

**Comment by capflam** [comment]
For me, it is a bug. `urlp` must use `'?'` by default as the first delimiter but not as parameters separator, except if it is explicitly specified. It is the way I understand the documentation:

> This extracts the first occurrence of the parameter <name> in the query
> string, which begins after either '?' or &lt;delim&gt;, and which ends before '&',
> ';' or &lt;delim&gt;.

If means that with your example , `urlp(x)` should return `y?z` while `url(x,?)` should return `y`.

**Comment by tmccombs** [comment]
Also, I forgot to mention, explicitly specifying a `<delim>` doesn't work, because it uses the `<delim>` for both the first delimiter and the paramter separator. 

I also wonder if it is worth having a more general `param` converter, with a signature like:

```
param(<name> [, <delim>])
```
that could be used like:

```
query,param(my-param)
```

and `<delim>` would default to `&`. I could make a PR for that if that is desirable? (and/or to change urlp to only use "?" for the first delimiter.

**Comment by wtarreau** [comment]
I like this approach. It's indeed much cleaner than what is being done. There's a very long history behind urlp, urlparam etc that justifies the current situation but maybe we should work on something to restart cleanly as you proposed here.

**Comment by wtarreau** [comment]
Just writing this down so that we don't lose it.

In parallel we've checked with Christopher and it turns out that adding an optional 3rd arg to urlp() to distinguish the query-string delimiter and the argument delimiter would also be nice to have and easy to implement (and that doesn't remove the value of your proposal).

**Comment by tmccombs** [comment]
There is a patch for the `param` converter in the mailing lists, but discussion on it kind of died: https://www.mail-archive.com/haproxy@formilux.org/msg42420.html
